Don't override user provided title and body when --fill is used

msfjarvis opened the pull request:
Fixes https://github.com/cli/cli/issues/1400

I'm not a Go developer and couldn't find any pre-existing tests for the `--fill` flag so haven't added any in this PR. Existing tests (ran using `make test`) still pass.

I created this PR using my generated binary with this command: `./bin/gh pr create -t "Don't override user provided title and body when --fill is used" --fill -d` so I guess that is a test in itself :D

msfjarvis committed:
Don't override user provided title and body when --fill is used

Signed-off-by: Harsh Shandilya <me@msfjarvis.dev>

bigcook12345 commented:
That's why I corrected my typo. No harm,  hopefully it helped out

mislav reviewed (changes requested):
Looks good!

However, if the git log could not be obtained, the `defaultsErr` condition would be printed and the operation would be aborted even when `--title` and `--body` were explicitly provided and thus the results of `git log` would not get used for anything.

The solution: only ever apply the logic of `Autofill` when `--title` and `--body` were not provided. What do you think?

msfjarvis commented:
> Looks good!
> 
> However, if the git log could not be obtained, the `defaultsErr` condition would be printed and the operation would be aborted even when `--title` and `--body` were explicitly provided and thus the results of `git log` would not get used for anything.
> 

That's true :(

> The solution: only ever apply the logic of `Autofill` when `--title` and `--body` were not provided. What do you think?

I'm not too keen on this idea, because it'd break the workflow I came here to fix. I have a couple hazy ideas on what I'd want this to do instead, and will update the PR once I'm back at a computer and able to form a clear path forward.

AliabbasMerchant commented:
@msfjarvis I would say be a bit careful when using the `body == ""` (instead use `cmd.Flags().Changed("body")`) condition, as the user may himself have specified the body to be an empty string.  
I believe in such cases(deciding when we need to select between our and user-given values), its best to use`cmd.Flags().Changed("someFlag")`

msfjarvis commented:
> @msfjarvis I would say be a bit careful when using the `body == ""` (instead use `cmd.Flags().Changed("body")`) condition, as the user may himself have specified the body to be an empty string.  
> I believe in such cases(deciding when we need to select between our and user-given values), its best to use`cmd.Flags().Changed("someFlag")`

Thanks for the tip!

msfjarvis commented:
This new revision works for all my needs and should also address your concerns @mislav. Unfortunately I ended up breaking the tests and don't know enough Go to fix them :(

msfjarvis commented on pkg/cmd/pr/create/create.go:
```diff
@@ -287,11 +289,15 @@ func createRun(opts *CreateOptions) error {
 		}
 	} else if opts.Autofill {
-		if defaultsErr != nil {
+		if defaultsErr != nil && !(titleProvided || bodyProvided) {
```
I'm not 100% sure this is 'correct'. When we fail to compute defaults, can we tell which of title or body failed to compute? If that's possible, then we could make this more precise and fail for cases such as

- couldn't compute default title && !titleProvided
- couldn't compute default body && !bodyProvided

msfjarvis commented on pkg/cmd/pr/create/create.go:
```diff
@@ -122,7 +122,7 @@ func NewCmdCreate(f *cmdutil.Factory, runF func(*CreateOptions) error) *cobra.Co
 }
 
-func createRun(opts *CreateOptions) error {
+func createRun(cmd *cobra.Command, opts *CreateOptions) error {
```
To be able to do `cmd.Flags().Changed(name string)` I ended up introducing an extra parameter, which is what broke the tests.

AliabbasMerchant replied:
@msfjarvis Add 2 more fields, `TitleProvided` and `BodyProvided` to `CreateOptions`, and access them via `opts` then

titleProvided and bodyProvided are calculated in lines 84-85 also

msfjarvis replied:
Thanks, I'll update this shortly

msfjarvis commented on pkg/cmd/pr/create/create.go:
```diff
@@ -78,7 +78,7 @@ func NewCmdCreate(f *cmdutil.Factory, runF func(*CreateOptions) error) *cobra.Co
 			$ gh pr create --project "Roadmap"
 			$ gh pr create --base develop --head monalisa:feature
-    	`),
+		`),
```
Small formatting nit that I came across, can revert if it's out of scope.

msfjarvis commented on pkg/cmd/pr/create/create.go:
```diff
@@ -287,11 +289,15 @@ func createRun(opts *CreateOptions) error {
 		}
-		title = defs.Title
-		body = defs.Body
+		if !titleProvided {
+			title = defs.Title
+		}
+		if !bodyProvided {
+			body = defs.Body
+		}
```
since we already set title and body to `opts.Title` and `opts.Body` respectively, I added these conditions to prevent overwriting them with defaults.

msfjarvis committed:
Refactor and fix tests

Signed-off-by: Harsh Shandilya <me@msfjarvis.dev>

mislav reviewed (approved):
Fantastic! Thank you ✨
